### PR TITLE
epoch trigger - sort metaheaders according to nonce.

### PR DESCRIPTION
**Pull request by sasurobert**
make it easier to search in local pool.
epoch trigger - subscribe to notifier
shardBlock - do not delete in add
metaBlock - do not delete in add

### CODE REVIEW

File: epochStart/shardchain/trigger.go [hunk 1]
```diff
@@ -283,42 +291,49 @@ func (t *trigger) ReceivedHeader(header data.HeaderHandler) {
...
 
+	t.mapHashHdr[string(hdrHash)] = metaHdr
+	t.mapNonceHashes[metaHdr.Nonce] = append(t.mapNonceHashes[metaHdr.Nonce], string(hdrHash))
+
+	sortedMetas := make([]metaInfo, 0, len(t.mapEpochStartHdrs))
 	for hash, meta := range t.mapEpochStartHdrs {
-		canActivateEpochStart, finalityAttestingRound := t.checkIfTriggerCanBeActivated(hash, meta)
-		if canActivateEpochStart && t.epoch < meta.Epoch {
-			t.epoch = meta.Epoch
+		metaInfo := metaInfo{
```
Comment (SebastianMarian) [inline epochStart/shardchain/trigger.go]: metaInfo := &metaInfo{

File: epochStart/shardchain/trigger.go [hunk 2]
```diff
@@ -283,42 +291,49 @@ func (t *trigger) ReceivedHeader(header data.HeaderHandler) {
...
-	} else {
-		t.mapHashHdr[string(hdrHash)] = metaHdr
-		t.mapNonceHashes[metaHdr.Nonce] = append(t.mapNonceHashes[metaHdr.Nonce], string(hdrHash))
 	}
 
+	t.mapHashHdr[string(hdrHash)] = metaHdr
+	t.mapNonceHashes[metaHdr.Nonce] = append(t.mapNonceHashes[metaHdr.Nonce], string(hdrHash))
+
+	sortedMetas := make([]metaInfo, 0, len(t.mapEpochStartHdrs))
 	for hash, meta := range t.mapEpochStartHdrs {
```
Comment (SebastianMarian) [inline epochStart/shardchain/trigger.go]: hash, hdr

File: process/block/metablock.go [hunk 3]
```diff
@@ -1506,6 +1496,10 @@ func (mp *metaProcessor) receivedShardHeader(headerHandler data.HeaderHandler, s
 		return
 	}
 
+	if !mp.blockTracker.ShouldAddHeader(shardHeader) {
```
Comment (SebastianMarian) [inline process/block/metablock.go]: Actually you can remove this condition at all as it was used only to delete the out of range received headers. The conditions which protect to not request miniblocks in advance is right above.

File: process/block/shardblock.go [hunk 4]
```diff
@@ -1359,6 +1348,10 @@ func (sp *shardProcessor) receivedMetaBlock(headerHandler data.HeaderHandler, me
 		return
 	}
 
+	if !sp.blockTracker.ShouldAddHeader(metaBlock) {
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: You can remove this check at all. The same situation as in metablock.go above

File: epochStart/shardchain/trigger.go [hunk 5]
```diff
@@ -261,16 +269,16 @@ func (t *trigger) setLowerEpochFinalityAttestingRoundIfNeeded(metaHdr *block.Met
 }
 
 // ReceivedHeader saves the header into pool to verify if end-of-epoch conditions are fulfilled only called with validated metablocks
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/shardchain/trigger.go]: done.

File: epochStart/shardchain/trigger.go [hunk 6]
```diff
@@ -283,42 +291,49 @@ func (t *trigger) ReceivedHeader(header data.HeaderHandler) {
...
-
-	} else {
-		t.mapHashHdr[string(hdrHash)] = metaHdr
-		t.mapNonceHashes[metaHdr.Nonce] = append(t.mapNonceHashes[metaHdr.Nonce], string(hdrHash))
 	}
 
+	t.mapHashHdr[string(hdrHash)] = metaHdr
+	t.mapNonceHashes[metaHdr.Nonce] = append(t.mapNonceHashes[metaHdr.Nonce], string(hdrHash))
+
+	sortedMetas := make([]metaInfo, 0, len(t.mapEpochStartHdrs))
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply epochStart/shardchain/trigger.go]: done

File: epochStart/shardchain/trigger.go (hunk 1 above)
Comment (sasurobert) [reply epochStart/shardchain/trigger.go]: done

File: epochStart/shardchain/trigger.go (hunk 2 above)
Comment (sasurobert) [reply epochStart/shardchain/trigger.go]: done

File: process/block/metablock.go (hunk 3 above)
Comment (sasurobert) [reply process/block/metablock.go]: done

File: process/block/shardblock.go (hunk 4 above)
Comment (sasurobert) [reply process/block/shardblock.go]: done

File: epochStart/shardchain/trigger.go [hunk 7]
```diff
@@ -283,42 +292,49 @@ func (t *trigger) ReceivedHeader(header data.HeaderHandler) {
...
 
-	for hash, meta := range t.mapEpochStartHdrs {
-		canActivateEpochStart, finalityAttestingRound := t.checkIfTriggerCanBeActivated(hash, meta)
-		if canActivateEpochStart && t.epoch < meta.Epoch {
-			t.epoch = meta.Epoch
+	sort.Slice(sortedMetaInfo, func(i, j int) bool {
+		return sortedMetaInfo[i].hdr.Epoch < sortedMetaInfo[j].hdr.Epoch
+	})
+
+	for _, metaInfo := range sortedMetaInfo {
```
Comment (ccorcoveanu) [inline epochStart/shardchain/trigger.go]: can you rename one of the `metaInfo` so the warning with shadow declaration is removed?

File: process/block/shardblock.go [hunk 8]
```diff
@@ -1349,7 +1339,6 @@ func (sp *shardProcessor) receivedMetaBlock(headerHandler data.HeaderHandler, me
 		return
 	}
 
-	sp.epochStartTrigger.ReceivedHeader(metaBlock)
 	if sp.epochStartTrigger.IsEpochStart() {
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: Here we can have a deadlock, as this channel "chRcvEpochStart" is not buffered and it is consumed only in the method requestEpochStartInfo called from ProcessBlock. This means that if node will not reach to that point all the metablocks received meantime will be stuck in this point, as now you will not call the sp.epochStartTrigger.ReceivedHeader(metaBlock) method synchronous, from here, anymore.
I will suggest to remove this channel at all and also these 3 lines from here, and to change the method requestEpochStartInfo like this:

func (sp *shardProcessor) requestEpochStartInfo(header *block.Header, haveTime func() time.Duration) error {
	haveMissingMetaHeaders := header.IsStartOfEpochBlock() && !sp.epochStartTrigger.IsEpochStart()
	if !haveMissingMetaHeaders {
		return nil
	}

	for {
		time.Sleep(time.Millisecond)
		
		if haveTime() < 0 {
			break
		}
		
		if sp.epochStartTrigger.IsEpochStart() {
			return nil
		}
	}

	return process.ErrTimeIsOut
}

Comment (iulianpascalau) [reply process/block/shardblock.go]: As an alternative, as to not have a blocking write, @sasurobert could have written something like:
```GO
select{
    case sp.chRcvEpochStart <- true:
    default:
}
```

**Pull request merged**